lib/std/mem.zig

```diff
@@ -876,7 +870,85 @@ pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, nee
     return null;
 }
 
+fn boyerMooreHorspoolPreprocessReverse(pattern: []const u8, table: *[256]usize) void {
+    for (table) |*c| {
+        c.* = pattern.len;
+    }
+
+    var i: usize = pattern.len - 1;
+    while (i > 0) : (i -= 1) {
+        table[pattern[i]] = i;
+    }
+}
+
+fn boyerMooreHorspoolPreprocess(pattern: []const u8, table: *[256]usize) void {
+    for (table) |*c| {
+        c.* = pattern.len;
+    }
+
+    var i: usize = 0;
+    while (i < pattern.len - 1) : (i += 1) {
+        table[pattern[i]] = pattern.len - 1 - i;
+    }
+}
+/// Find the index in a slice of a sub-slice, searching from the end backwards.
+/// To start looking at a different index, slice the haystack first.
+// Reverse boyer-moore-horspool algorithm
+pub fn lastIndexOf(comptime T: type, haystack: []const T, needle: []const T) ?usize {
+    if (needle.len > haystack.len or needle.len == 0) return null;
+
+    // byte comparison with float doesn't work because +0.0 and -0.0 and considered
+    // equal but their byte representations are not equal.
+    if (@typeInfo(T) == .Float or haystack.len < 32 or needle.len <= 2)
+        return lastIndexOfLinear(T, haystack, needle);
+
+    const haystack_bytes = sliceAsBytes(haystack);
+    const needle_bytes = sliceAsBytes(needle);
+
+    var skip_table: [256]usize = undefined;
+    boyerMooreHorspoolPreprocessReverse(needle_bytes, skip_table[0..]);
+
+    var i: usize = haystack_bytes.len - needle_bytes.len;
+    while (true) {
+        if (mem.eql(u8, haystack_bytes[i .. i + needle_bytes.len], needle_bytes)) return i;
+        const skip = skip_table[haystack_bytes[i]];
+        if (skip > i) break;
+        i -= skip;
+    }
+
+    return null;
+}
+
+// Boyer-moore-horspool algorithm
+pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, needle: []const T) ?usize {
+    if (needle.len > haystack.len or needle.len == 0) return null;
+
+    // byte comparison with float doesn't work because +0.0 and -0.0 and considered
+    // equal but their byte representations are not equal.
+    if (@typeInfo(T) == .Float or haystack.len < 32 or needle.len <= 2)
```

Comment:
Hmm, maybe `std.traits.hasUniqueRepresentation(T)` can be used instead.
It has an interesting comment for integer types: `TODO check that it is still true`. that may not be true for non-multiple-of-8 types.

Reply:
Looks like that would work, fixed